File: frontend/src/types/operations.ts
/**
 * Monitoring, scenario and settings response types.
 *
 * Note the shape of every monitoring measure: an `available`/`computable` flag
 * plus a `reason`. The UI must render "not yet computable" differently from
 * "measured, and fine" — they are different answers, and a page that showed
 * zero for both would be asserting a check that never ran.
 */

export interface Freshness {
  demand_history_end: string | null;
  demand_history_age_months: number | null;
  stock_snapshot_date: string;
  stock_snapshot_age_months: number | null;
  panel_built_at: string | null;
  panel_rows: number;
  dataset_ingested_at: string | null;
  note: string;
}

export interface DriftWindow {
  months: number;
  mean_monthly_demand: number | null;
  rows: number;
}

export interface DriftRow {
  scope_level: string;
  scope_key: string;
  earlier_mean_monthly_demand: number | null;
  recent_mean_monthly_demand: number | null;
  shift_pct: number | null;
}

export interface Drift {
  available: boolean;
  reason: string | null;
  window_months: number;
  recent_periods: string[];
  national: {
    earlier: DriftWindow;
    recent: DriftWindow;
    shift_pct: number | null;
  } | null;
  target_sources_recent: string[];
  target_sources_earlier: string[];
  target_source_changed: boolean;
  rows: DriftRow[];
  note: string | null;
}

export interface ChampionAgeRow {
  scope_kind: string;
  scope_key: string;
  champion_model_id: string;
  selection_source: string;
  selected_at: string | null;
  age_days: number | null;
  training_run_id: string;
  from_newest_training_run: boolean;
  beaten_by_baseline: boolean;
}

export interface ChampionAge {
  active_champions: number;
  newest_training_run_id: string | null;
  champions_from_older_runs: number;
  champions_beaten_by_a_baseline: number;
  rows: ChampionAgeRow[];
  note: string;
}

export interface DeteriorationRow {
  scope_level: string;
  scope_key: string;
  period: string;
  model_id: string | null;
  actual: number;
  point_forecast: number;
  absolute_error: number;
  accrued_wape: number | null;
  backtest_wape: number | null;
  deterioration_pct: number | null;
}

export interface Deterioration {
  /** False means it has not been checked — not that nothing deteriorated. */
  computable: boolean;
  reason: string | null;
  origin_period: string | null;
  forecast_periods: string[];
  observed_forecast_periods: string[];
  rows: DeteriorationRow[];
  note: string | null;
}

export interface MonitoringResponse {
  generated_at: string;
  freshness: Freshness;
  drift: Drift;
  champions: ChampionAge;
  deterioration: Deterioration;
  note: string;
}

export interface ScenarioRequest {
  name?: string;
  baseline_forecast_run_id?: string | null;
  demand_multiplier?: number;
  service_level?: number;
  lead_time_days?: number | null;
  review_period_days?: number | null;
  scope_level?: string;
  scope_keys?: string[] | null;
  period?: string | null;
  limit?: number;
}

export interface ScenarioRow {
  scope_key: string;
  period: string;
  horizon: number;
  model_id: string | null;
  baseline_point: number | null;
  scenario_point: number | null;
  baseline_quantile: number | null;
  scenario_quantile: number | null;
  baseline_recommended_order: number | null;
  scenario_recommended_order: number | null;
  unavailable_reason: string | null;
}

export interface ScenarioResponse {
  name: string;
  generated_at: string;
  baseline_forecast_run_id: string;
  baseline_origin_period: string;
  scope_level: string;
  period: string | null;
  levers: {
    demand_multiplier: number;
    service_level: number;
    lead_time_days: number | null;
    baseline_lead_time_days: number;
    review_period_days: number;
    baseline_review_period_days: number;
  };
  totals: {
    baseline_demand: number;
    scenario_demand: number;
    demand_delta: number;
    demand_delta_pct: number | null;
    baseline_order_quantity: number;
    scenario_order_quantity: number;
    order_delta: number;
    order_delta_pct: number | null;
  };
  rows: ScenarioRow[];
  rows_returned: number;
  rows_without_baseline: number;
  caveats: string[];
}

export interface SettingsResponse {
  app_name: string;
  app_version: string;
  environment: string;
  official_model_count: number;
  registered_model_ids: string[];
  baseline_method_ids: string[];
  min_history_profile: string;
  xgboost_training_profile: string;
  forecast_horizon_months: number;
  service_levels: number[];
  random_seed: number;
  max_local_series: number;
  local_series_selection: string;
  per_model_timeout_seconds: number;
  lstm_timeout_seconds: number;
  max_training_workers: number;
  review_period_days: number;
  default_lead_time_days: number;
  stock_snapshot_date: string;
  mlflow_enabled: boolean;
  mlflow_tracking_uri: string;
  database_dialect: string;
  expected_controls: Record<string, number>;
  notes: string[];
}

File: backend/app/schemas/operations.py
"""Monitoring, scenario and settings payloads.

Every monitoring measure is `available`/`computable` plus a `reason`, never a
bare number that might be a default. "Not yet computable" and "no problem
detected" are different answers, and the schema forces the distinction.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import Field, field_validator

from app.domain.ais.inventory import SERVICE_LEVELS
from app.schemas.common import ApiModel


class FreshnessOut(ApiModel):
    demand_history_end: str | None = None
    demand_history_age_months: int | None = None
    stock_snapshot_date: str
    stock_snapshot_age_months: int | None = None
    panel_built_at: datetime | None = None
    panel_rows: int = 0
    dataset_ingested_at: datetime | None = None
    note: str


class DriftRow(ApiModel):
    scope_level: str
    scope_key: str
    earlier_mean_monthly_demand: float | None = None
    recent_mean_monthly_demand: float | None = None
    shift_pct: float | None = None


class DriftWindow(ApiModel):
    months: int
    mean_monthly_demand: float | None = None
    rows: int


class DriftComparison(ApiModel):
    earlier: DriftWindow
    recent: DriftWindow
    shift_pct: float | None = None


class DriftOut(ApiModel):
    available: bool
    reason: str | None = None
    window_months: int
    recent_periods: list[str] = Field(default_factory=list)
    national: DriftComparison | None = None
    target_sources_recent: list[str] = Field(default_factory=list)
    target_sources_earlier: list[str] = Field(default_factory=list)
    target_source_changed: bool = False
    rows: list[DriftRow] = Field(default_factory=list)
    note: str | None = None


class ChampionAgeRow(ApiModel):
    scope_kind: str
    scope_key: str
    champion_model_id: str
    selection_source: str
    selected_at: datetime | None = None
    age_days: float | None = None
    training_run_id: str
    from_newest_training_run: bool
    beaten_by_baseline: bool


class ChampionAgeOut(ApiModel):
    active_champions: int
    newest_training_run_id: str | None = None
    champions_from_older_runs: int
    champions_beaten_by_a_baseline: int
    rows: list[ChampionAgeRow] = Field(default_factory=list)
    note: str


class DeteriorationRow(ApiModel):
    scope_level: str
    scope_key: str
    period: str
    model_id: str | None = None
    actual: float
    point_forecast: float
    absolute_error: float
    accrued_wape: float | None = None
    backtest_wape: float | None = None
    deterioration_pct: float | None = None


class DeteriorationOut(ApiModel):
    #: False means it has not been checked, which is not the same as "no
    #: deterioration". The reason says which.
    computable: bool
    reason: str | None = None
    origin_period: str | None = None
    forecast_periods: list[str] = Field(default_factory=list)
    observed_forecast_periods: list[str] = Field(default_factory=list)
    rows: list[DeteriorationRow] = Field(default_factory=list)
    note: str | None = None


class MonitoringResponse(ApiModel):
    generated_at: datetime
    #: What this deployment reports on. Carried so a two-branch figure can
    #: never read as a national one - the restriction is applied upstream, and
    #: this is what states it (docs/DECISIONS.md D-049).
    workspace_scope: dict[str, Any] | None = None
    freshness: FreshnessOut
    drift: DriftOut
    champions: ChampionAgeOut
    deterioration: DeteriorationOut
    note: str


class ScenarioRequest(ApiModel):
    name: str = Field(default="Scenario", max_length=120)
    baseline_forecast_run_id: str | None = Field(
        default=None, description="Defaults to the newest completed forecast run."
    )
    demand_multiplier: float = Field(default=1.0, gt=0)
    service_level: int = 95
    lead_time_days: float | None = Field(default=None, ge=0)
    review_period_days: int | None = Field(default=None, ge=1)
    scope_level: str = "branch"
    scope_keys: list[str] | None = None
    period: str | None = None
    limit: int = Field(default=200, ge=1, le=5000)

    @field_validator("service_level")
    @classmethod
    def _known_level(cls, value: int) -> int:
        if value not in SERVICE_LEVELS:
            raise ValueError(
                f"Service level {value} is not offered. Valid: {list(SERVICE_LEVELS)}."
            )
        return value


class ScenarioRow(ApiModel):
    scope_key: str
    period: str
    horizon: int
    model_id: str | None = None
    baseline_point: float | None = None
    scenario_point: float | None = None
    baseline_quantile: float | None = None
    scenario_quantile: float | None = None
    baseline_recommended_order: float | None = None
    scenario_recommended_order: float | None = None
    unavailable_reason: str | None = None


class ScenarioLevers(ApiModel):
    demand_multiplier: float
    service_level: int
    lead_time_days: float | None = None
    baseline_lead_time_days: float
    review_period_days: int
    baseline_review_period_days: int


class ScenarioTotals(ApiModel):
    baseline_demand: float
    scenario_demand: float
    demand_delta: float
    demand_delta_pct: float | None = None
    baseline_order_quantity: float
    scenario_order_quantity: float
    order_delta: float
    order_delta_pct: float | None = None


class ScenarioResponse(ApiModel):
    name: str
    generated_at: datetime
    baseline_forecast_run_id: str
    baseline_origin_period: str
    scope_level: str
    period: str | None = None
    levers: ScenarioLevers
    totals: ScenarioTotals
    rows: list[ScenarioRow]
    rows_returned: int
    rows_without_baseline: int
    caveats: list[str]


class SettingsResponse(ApiModel):
    app_name: str
    app_version: str
    environment: str
    official_model_count: int
    registered_model_ids: list[str]
    baseline_method_ids: list[str]
    min_history_profile: str
    xgboost_training_profile: str
    forecast_horizon_months: int
    service_levels: list[int]
    random_seed: int
    max_local_series: int
    local_series_selection: str
    per_model_timeout_seconds: float
    lstm_timeout_seconds: float
    max_training_workers: int
    review_period_days: int
    default_lead_time_days: float
    stock_snapshot_date: str
    mlflow_enabled: bool
    mlflow_tracking_uri: str
    #: Dialect only. A settings page is not a place to publish a connection
    #: string or a filesystem layout.
    database_dialect: str
    expected_controls: dict[str, int]
    notes: list[str]
